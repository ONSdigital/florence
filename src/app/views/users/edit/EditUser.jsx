import React, { useEffect, useCallback, useState } from "react";
import isEqual from "lodash/isEqual";
import PropTypes from "prop-types";
import { Link } from "react-router";
import url from "../../../utilities/url";
import Input from "../../../components/Input";
import Warning from "../../../icons/Warning";
import validate from "./validate";
import FormValidationError from "./ValidationErrors";
import BackButton from "../../../components/back-button/BackButton";
import FormFooter from "../../../components/form-footer";
import SelectedItemList from "../../../components/selected-items/SelectedItemList";
import Loader from "../../../components/loader/index";
import RadioGroup from "../../../components/radio-buttons/RadioGroup";
import TextArea from "../../../components/text-area/TextArea";

const options = [
    {
        id: "active",
        value: "active",
        label: "Active",
    },
    {
        id: "suspended",
        value: "suspended",
        label: "Suspended",
    },
];

const EditUser = props => {
    const id = props.params.userID;

    useEffect(() => {
        if (id) {
            props.loadUser(id);
        }
    }, [id]);

    const { user, userGroups, loading, updateUser } = props;
    const [values, setValues] = useState(null);
    const [errors, setErrors] = useState({});
    const [isSubmitting, setIsSubmitting] = useState(false);
    const hasErrors = errors && Object.keys(errors).length > 0;
    const hasValues = !isEqual(values, user);

    useEffect(() => {
        if (user) {
            setValues({ ...user });
        }
    }, [user]);

    const handleSubmit = e => {
        if (e) e.preventDefault();
        setErrors(validate(values));
        setIsSubmitting(true);
        if (hasErrors) return;
        updateUser(id, values);
    };

    const handleChange = e => {
        const key = e.target.name;
        const val = e.target.value;
        setIsSubmitting(false);

        if (Object.keys(errors).length > 0) {
            delete errors[e.target.name];
            setErrors(errors);
        }
        setValues(values => ({ ...values, [key]: val }));
    };

    const handleAccessChange = e => {
        setValues(values => ({ ...values, active: e.id === "active" }));
    };

    if (loading) return <Loader classNames="grid grid--align-center grid--align-self-center grid--full-height" />;

    if (!user) return <h1>No user found.</h1>;

    return (
        <form className="form">
            <div className="grid grid--justify-space-around">
                <div className="grid__col-11 grid__col-md-9">
                    <BackButton classNames={"margin-top--2"} />
                    <div className="grid grid--justify-space-between">
                        <div className="grid__col-md-6">
                            <h1 className="margin-top--1 margin-bottom--1">{`${user.forename} ${user.lastname}`}</h1>
                            <p>{user.id}</p>
                            {hasErrors && <FormValidationError errors={errors} />}
                            <div className="grid">
                                <div className="grid__col-lg-6 margin-top--1">
                                    <Input
                                        id="forename"
                                        name="forename"
                                        label="First name"
                                        type="text"
                                        error={errors?.forename}
                                        value={values?.forename ? values.forename : ""}
                                        onChange={handleChange}
                                    />
                                    <Input
                                        id="lastname"
                                        name="lastname"
                                        label="Last name"
                                        type="text"
                                        error={errors?.lastname}
                                        onChange={handleChange}
                                        value={values?.lastname ? values.lastname : ""}
                                    />
                                    <h2 className="margin-top--1">Team Member</h2>
                                    {!userGroups.length == 0 && <p>User is not a member of a team. Please add them to a team.</p>}
                                    {userGroups && (
                                        <SelectedItemList
                                            classNames="selected-item-list__item--info"
                                            items={userGroups.map(gr => ({ id: gr, name: gr }))}
                                        />
                                    )}
                                </div>
                            </div>
                        </div>
                        <div className="grid__col-md-5">
                            <div className="form-group margin-top--1 margin-bottom--1">
                                <RadioGroup
                                    groupName="active"
                                    radioData={options}
                                    selectedValue={values?.active ? "active" : "suspended"}
                                    onChange={handleAccessChange}
                                    legend="User Access"
                                    inline
                                />
                                <Input
                                    id="status_notes"
                                    label="Notes"
                                    name="status_notes"
                                    type="textarea"
                                    value={values?.status_notes}
                                    onChange={handleChange}
                                />
                                <p>
                                    Add the reason for changing the users' access. For example, they have left or no longer need access to Florence.
                                    Do not include any personal information.
                                </p>
                            </div>
                        </div>
                    </div>
                </div>
                <FormFooter hasValues={hasValues} loading={loading} handleSubmit={handleSubmit} />
            </div>
        </form>
    );
};

export default EditUser;
