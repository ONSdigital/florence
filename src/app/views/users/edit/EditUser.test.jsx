import React from "react";
import { render, screen, within } from "@testing-library/react";
import renderer from "react-test-renderer";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom/extend-expect";
import EditUser from "./EditUser";
import { groups, user } from "../../../utilities/tests/mockData";
import { createMockUser } from "../../../utilities/tests/test-utils";

const admin = createMockUser("admin@test.com", true, true, "ADMIN");
const editor = createMockUser("editor@test.com", false, true, "EDITOR");

const props = {
    loading: false,
    params: { id: "test.user-1498@ons.gov.uk" },
    rootPath: "test",
    user: user,
    userGroups: [],
    updateUser: jest.fn(),
    loadUser: jest.fn(),
    loadUserGroups: jest.fn(),
    router: { setRouteLeaveHook: jest.fn() },
    loggedInUser: admin,
};
const setRouteLeaveHook = jest.fn();

describe("EditUser", () => {
    it("matches the snapshot", () => {
        const tree = renderer.create(
            <EditUser.WrappedComponent {...props} params={{ id: "test.user-1498@ons.gov.uk", router: setRouteLeaveHook }} />
        );
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it("fetches user details on load", () => {
        render(<EditUser.WrappedComponent {...props} params={{ id: "test.user-1498@ons.gov.uk", router: setRouteLeaveHook }} />);
        expect(props.loadUser).toBeCalledWith("test.user-1498@ons.gov.uk");
        expect(props.loadUserGroups).toBeCalled();
    });

    it("shows the form with user data", () => {
        render(<EditUser.WrappedComponent {...props} params={{ id: "test.user-1498@ons.gov.uk", router: setRouteLeaveHook }} />);

        expect(screen.getByRole("link", { name: "Back" })).toBeInTheDocument();
        expect(screen.getByRole("heading", { level: 1 })).toHaveTextContent("test user-1498");
        expect(screen.getByText("test.user-1498@ons.gov.uk")).toBeInTheDocument();

        expect(screen.getByLabelText(/First name/i)).toHaveValue("test");
        expect(screen.getByLabelText(/Last name/i)).toHaveValue("user-1498");

        expect(screen.getByRole("group", { name: /User Access/i })).toBeInTheDocument();
        expect(screen.getByRole("radio", { name: /Active/i })).toBeChecked();
        expect(screen.getByRole("radio", { name: /Suspended/i })).not.toBeChecked();
        expect(screen.getByLabelText(/Notes/i)).toHaveValue("This user is active");

        expect(screen.getByRole("button", { name: /Save changes/i })).toBeInTheDocument();
        expect(screen.getByRole("button", { name: /Cancel/i })).toBeInTheDocument();
        expect(screen.queryByText(/You have unsaved changes/i)).not.toBeInTheDocument();
    });

    it("allows editing and shows unsaved changes message", () => {
        render(<EditUser.WrappedComponent {...props} params={{ id: "test.user-1498@ons.gov.uk", router: setRouteLeaveHook }} />);

        userEvent.clear(screen.getByLabelText(/First name/i));
        userEvent.paste(screen.getByLabelText(/First name/i), "My test First name");

        expect(screen.getByLabelText(/First name/i)).toHaveValue("My test First name");
        expect(screen.getByText(/You have unsaved changes/i)).toBeInTheDocument();
    });

    it("validates form and display errors in panel and within input and disables the submit button", () => {
        render(<EditUser.WrappedComponent {...props} params={{ id: "test.user-1498@ons.gov.uk", router: setRouteLeaveHook }} />);

        expect(screen.getByLabelText(/First name/i)).toHaveValue("test");
        expect(screen.getByLabelText(/Last name/i)).toHaveValue("user-1498");

        userEvent.clear(screen.getByLabelText(/First name/i));
        userEvent.clear(screen.getByLabelText(/Last name/i));

        expect(screen.getByLabelText(/First name/i)).toHaveValue("");
        expect(screen.getByLabelText(/Last name/i)).toHaveValue("");

        userEvent.click(screen.getByText(/save changes/i));

        expect(screen.getByText(/Fix the following:/i)).toBeInTheDocument();
        expect(screen.getByRole("link", { name: /Please enter a first name/i })).toBeInTheDocument();
        expect(screen.getByRole("link", { name: /Please enter a last name/i })).toBeInTheDocument();

        within(screen.getByTestId("forename")).getByText(/Please enter a first name/i);
        within(screen.getByTestId("lastname")).getByText(/Please enter a last name/i);

        expect(screen.getByRole("button", { name: /save changes/i })).toBeDisabled();
    });

    it("shows loader when fetching user data", () => {
        const newProps = {
            ...props,
            loading: true,
        };
        render(<EditUser.WrappedComponent {...newProps} params={{ id: "test.user-1498@ons.gov.uk", router: setRouteLeaveHook }} />);
        expect(screen.getByTestId("loader")).toBeInTheDocument();
    });

    it("updates user data", async () => {
        render(<EditUser.WrappedComponent {...props} params={{ id: "test.user-1498@ons.gov.uk", router: setRouteLeaveHook }} />);

        userEvent.paste(screen.getByLabelText(/First name/i), "boo");

        expect(screen.getByLabelText(/First name/i)).toHaveValue("testboo");
        expect(screen.getByText(/You have unsaved changes/i)).toBeInTheDocument();

        await userEvent.click(screen.getByText(/save changes/i));

        expect(props.updateUser).toBeCalledWith("test.user-1498@ons.gov.uk", {
            active: true,
            email: "test.user-1498@ons.gov.uk",
            forename: "testboo",
            groups: [],
            id: "test.user-1498@ons.gov.uk",
            lastname: "user-1498",
            status: "CONFIRMED",
            status_notes: "This user is active",
        });
    });
    it("shows groups user belongs to", () => {
        const newProps = {
            ...props,
            userGroups: groups,
        };

        render(<EditUser.WrappedComponent {...newProps} params={{ id: "test.user-1498@ons.gov.uk", router: setRouteLeaveHook }} />);

        expect(screen.getByText(/my test group description/i)).toBeInTheDocument();
        expect(screen.getByText(/my first test group description/i)).toBeInTheDocument();
        expect(screen.getByText(/admins group description/i)).toBeInTheDocument();
    });
});

describe("EditUser without admin permissions", () => {
    const editorProps = {
        ...props,
        loggedInUser: editor,
    };

    it("matches the snapshot", () => {
        const tree = renderer.create(
            <EditUser.WrappedComponent {...editorProps} params={{ id: "test.user-1498@ons.gov.uk", router: setRouteLeaveHook }} />
        );
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it("fetches user details on load", () => {
        render(<EditUser.WrappedComponent {...editorProps} params={{ id: "test.user-1498@ons.gov.uk", router: setRouteLeaveHook }} />);
        expect(editorProps.loadUser).toBeCalledWith("test.user-1498@ons.gov.uk");
        expect(editorProps.loadUserGroups).toBeCalled();
    });

    it("shows the form with user data and no admin fields", () => {
        render(<EditUser.WrappedComponent {...editorProps} params={{ id: "test.user-1498@ons.gov.uk", router: setRouteLeaveHook }} />);

        expect(screen.getByRole("link", { name: "Back" })).toBeInTheDocument();
        expect(screen.getByRole("heading", { level: 1 })).toHaveTextContent("test user-1498");
        expect(screen.getByText("test.user-1498@ons.gov.uk")).toBeInTheDocument();

        expect(screen.queryByLabelText(/First name/i)).not.toBeInTheDocument();
        expect(screen.queryByLabelText(/Last name/i)).not.toBeInTheDocument();

        expect(screen.queryByRole("group", { name: /User Access/i })).not.toBeInTheDocument();
        expect(screen.queryByRole("radio", { name: /Active/i })).not.toBeInTheDocument();
        expect(screen.queryByRole("radio", { name: /Suspended/i })).not.toBeInTheDocument();
        expect(screen.queryByLabelText(/Notes/i)).not.toBeInTheDocument();

        expect(screen.queryByRole("button", { name: /Save changes/i })).not.toBeInTheDocument();
        expect(screen.queryByRole("button", { name: /Cancel/i })).not.toBeInTheDocument();
        expect(screen.queryByText(/You have unsaved changes/i)).not.toBeInTheDocument();
    });
    it("shows loader when fetching user data", () => {
        const newProps = {
            ...editorProps,
            loading: true,
        };
        render(<EditUser.WrappedComponent {...newProps} params={{ id: "test.user-1498@ons.gov.uk", router: setRouteLeaveHook }} />);
        expect(screen.getByTestId("loader")).toBeInTheDocument();
    });
    it("shows groups user belongs to", () => {
        const newProps = {
            ...editorProps,
            userGroups: groups,
        };

        render(<EditUser.WrappedComponent {...newProps} params={{ id: "test.user-1498@ons.gov.uk", router: setRouteLeaveHook }} />);

        expect(screen.getByText(/my test group description/i)).toBeInTheDocument();
        expect(screen.getByText(/my first test group description/i)).toBeInTheDocument();
        expect(screen.getByText(/admins group description/i)).toBeInTheDocument();
    });
});
