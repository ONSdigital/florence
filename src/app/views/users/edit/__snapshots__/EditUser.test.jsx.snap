// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditUser matches the snapshot 1`] = `
<form
  className="form"
>
  <div
    className="grid grid--justify-space-around"
  >
    <div
      className="grid__col-11 grid__col-md-9"
    >
      <div
        className="margin-top--2"
      >
        ◀ 
        <a
          onClick={[Function]}
          role="link"
          style={Object {}}
        >
          Back
        </a>
      </div>
      <div
        className="grid grid--justify-space-between"
      >
        <div
          className="grid__col-md-6"
        >
          <h1
            className="margin-top--1 margin-bottom--1"
          >
            test user-1498
          </h1>
          <p>
            test.user-1498@ons.gov.uk
          </p>
          <div
            className="grid"
          >
            <div
              className="grid__col-lg-6 margin-top--1"
            >
              <div
                className="form__input"
              >
                <label
                  className="form__label"
                  htmlFor="forename"
                >
                  First name
                </label>
                <input
                  autoComplete="new-password"
                  autoFocus={false}
                  className="input"
                  disabled={false}
                  id="forename"
                  name="forename"
                  onChange={[Function]}
                  onFocus={[Function]}
                  type="text"
                  value=""
                />
              </div>
              <div
                className="form__input"
              >
                <label
                  className="form__label"
                  htmlFor="lastname"
                >
                  Last name
                </label>
                <input
                  autoComplete="new-password"
                  autoFocus={false}
                  className="input"
                  disabled={false}
                  id="lastname"
                  name="lastname"
                  onChange={[Function]}
                  onFocus={[Function]}
                  type="text"
                  value=""
                />
              </div>
              <h2
                className="margin-top--1"
              >
                Team Member
              </h2>
              <p>
                User is not a member of a team. Please add them to a team.
              </p>
            </div>
          </div>
        </div>
        <div
          className="grid__col-md-5"
        >
          <div
            className="form-group margin-top--1 margin-bottom--1"
          >
            <fieldset
              className="fieldset"
            >
              <legend
                className="fieldset__legend"
              >
                User Access
              </legend>
              <div
                className="radio radio--inline"
              >
                <input
                  checked={false}
                  className="radio__input"
                  id="active"
                  name="active"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  type="radio"
                  value="active"
                />
                <label
                  className="radio__label"
                  htmlFor="active"
                >
                  Active
                   
                </label>
              </div>
              <div
                className="radio radio--inline"
              >
                <input
                  checked={true}
                  className="radio__input selected"
                  id="suspended"
                  name="active"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  type="radio"
                  value="suspended"
                />
                <label
                  className="radio__label radio__label--selected"
                  htmlFor="suspended"
                >
                  Suspended
                   
                </label>
              </div>
            </fieldset>
            <div
              className="form__input"
            >
              <label
                className="form__label"
                htmlFor="status_notes"
              >
                Notes
              </label>
              <textarea
                autoFocus={false}
                className="input input__textarea"
                disabled={false}
                id="status_notes"
                name="status_notes"
                onChange={[Function]}
              />
            </div>
            <p>
              Add the reason for changing the users' access. For example, they have left or no longer need access to Florence. Do not include any personal information.
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      className="grid grid--justify-space-around padding-bottom--1 padding-top--1 form-footer"
      data-testid="form-footer"
    >
      <div
        className="grid__col-9"
      >
        <div
          className="grid grid--align-baseline"
        >
          <button
            className="btn btn--positive btn--margin-right"
            disabled={false}
            onClick={[Function]}
            type="submit"
          >
            Save changes
          </button>
          <a
            className="btn btn--invert-primary btn--margin-right"
            onClick={[Function]}
            role="button"
            style={Object {}}
          >
            Cancel
          </a>
          <span>
            <svg
              aria-label="warning icon"
              className="svg-icon--action-bar"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256,34.297L0,477.703h512L256,34.297z M256,422.05c-9.22,0-16.696-7.475-16.696-16.696s7.475-16.696,16.696-16.696 c9.22,0,16.696,7.475,16.696,16.696S265.22,422.05,256,422.05z M239.304,344.137V177.181h33.391v166.956H239.304z"
              />
            </svg>
             You have unsaved changes
          </span>
        </div>
      </div>
    </div>
  </div>
</form>
`;
